### PR TITLE
Joy axis update

### PR DESCRIPTION
**Pull request by Markil3**
https://hub.jmonkeyengine.org/t/joystickcompatibilitymappings-overhaul/44337
This PR adds two improvements to the JoystickCompatabilityMapping class: First of all, it allows the remapping file to specify whether a mapping is a buttom remapping or an axis remapping. Considering some gamepads will use the same name for a button and an axis, this feature will help in this field. It is an optional, backwards-compatible feature that can be enabled by adding "button." or "axis." in front of the property name.
The second change is the ability to change the effective range of axes by adding a bracketed tupple of two floats to the value
`axis.Xbox\ 360\ Wireless\ Receiver.z=rx[0.0,1.0]`
Note that the "axis." prefix is optional, and using the range remapper implies that this remapping is for axes.

### CODE REVIEW

**Comment by stephengold** [comment]
Thanks for contributing!

File: jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java [hunk 1]
```diff
@@ -54,16 +54,24 @@
...
  *
  *  @author    Paul Speed
+ *  @author Markil3
  */
 public class JoystickCompatibilityMappings {
 
     private static final Logger logger = Logger.getLogger(JoystickCompatibilityMappings.class.getName());
 
+    static {
+        logger.setLevel(Level.FINE);
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java:64]: FINE logging must be removed before integration.

File: jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java [hunk 2]
```diff
@@ -79,9 +87,175 @@
...
         }
-        return result;          
+        return result;
     }
- 
+
+    /**
+     * Obtains mappings specific to the joystick axis
+     * @param joystickName
+     * @param create
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java:96]: Please document the semantics of this parameter.

File: jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java [hunk 3]
```diff
@@ -79,9 +87,175 @@
...
+            result = new HashMap<String,AxisData>();
+            axisMappings.put(joystickName.trim(),result);
+        }
+        return result;
+    }
+
+    /**
+     * Obtains mappings specific to the joystick axis
+     * @param joystickName
+     * @param create
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java:112]: Please document the semantics of this parameter.

File: jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java [hunk 4]
```diff
@@ -79,9 +87,175 @@
...
+    protected static Map<String,String> getButtonMappings( String joystickName, boolean create ) {
+        Map<String,String> result = buttonMappings.get(joystickName.trim());
+        if ( result == null && create ) {
+            result = new HashMap<String,String>();
+            buttonMappings.put(joystickName.trim(),result);
+        }
+        return result;
+    }
+
+    /**
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java:125]: Please document what this method does.

File: jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java [hunk 5]
```diff
@@ -79,9 +87,175 @@
...
+         * need to go searching for it every tick), but that there is no remapping.
+         */
+        if (range.length == 0) {
+            return currentValue;
+        }
+
+        return (currentValue + range[1] + range[0]) * ((range[1] - range[0]) / 2);
+    }
+
+    /**
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java:168]: Please document what this method does.

File: jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java [hunk 6]
```diff
@@ -79,9 +87,175 @@
...
+        if (map != null && map.containsKey(componentId)) {
+            logger.log(Level.FINE, "returning remapped:" + map.get(componentId));
+            return ((String) map.get(componentId));
+        }
+
+        return componentId;
+    }
+
+    /**
+     * @param joystickName
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java:214]: Please document what this method does.

File: jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java [hunk 7]
```diff
@@ -79,9 +87,175 @@
...
+     * @param componentId
+     * @return
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/input/JoystickCompatibilityMappings.java:216]: Please document the return value.

File: jme3-core/src/main/java/com/jme3/input/event/JoyAxisEvent.java [hunk 8]
```diff
@@ -36,17 +36,23 @@
...
 
     private JoystickAxis axis;
     private float value;
+    private float rawValue;
 
     public JoyAxisEvent(JoystickAxis axis, float value) {
+        this(axis, value, value);
+    }
+
+    public JoyAxisEvent(JoystickAxis axis, float value, float rawValue) {
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/input/event/JoyAxisEvent.java:52]: Please document what this constructor does.

Comment (stephengold) [reply jme3-core/src/main/java/com/jme3/input/event/JoyAxisEvent.java:65]: Great! Thank you.

File: jme3-lwjgl/src/main/java/com/jme3/input/lwjgl/JInputJoyInput.java [hunk 9]
```diff
@@ -1,5 +1,9 @@
 /*
+<<<<<<< HEAD
+ * Copyright (c) 2009-2016 jMonkeyEngine
+=======
```
Comment (stephengold) [inline jme3-lwjgl/src/main/java/com/jme3/input/lwjgl/JInputJoyInput.java:4]: please resolve the edit conflict

File: jme3-lwjgl/src/main/java/com/jme3/input/lwjgl/JInputJoyInput.java [hunk 10]
```diff
@@ -1,5 +1,9 @@
...
  * Copyright (c) 2009-2021 jMonkeyEngine
+>>>>>>> master
```
Comment (stephengold) [inline jme3-lwjgl/src/main/java/com/jme3/input/lwjgl/JInputJoyInput.java:6]: please delete this line

**Review by stephengold — APPROVED** [review]
Next time, please don't reformat the code (unless that's the main purpose of the PR). It makes code review unnecessarily difficult.

**Comment by Markil3** [comment]
Right. Sorry for the pain. I still haven't figured out entirely how to get Android Studio to match your formatting standards entirely.

**Comment by stephengold** [comment]
Getting Android Studio to match our formatting standards isn't the issue here. The standards are relatively recent, and most existing code doesn't meet those standards.

The key thing for PRs like this is to prevent Android Studio from reformatting existing code!

**Comment by stephengold** [comment]
Unless there's substantial discussion of this PR, I plan to integrate it in about 8 hours.

**Pull request merged**